### PR TITLE
feat: permissions awareness (part 1)

### PR DESCRIPTION
**Pull request by davidlougheed**
some basic UI permissions awareness in place. lots more fine-grain work + dataset permissions + UI for grants/groups needed.

### CODE REVIEW

File: src/components/RequireAuth.js
Comment (v-rocheleau) [inline src/components/RequireAuth.js:31]: neat

File: src/components/AutoAuthenticate.js
Comment (v-rocheleau) [inline src/components/AutoAuthenticate.js:1]: Make new files .ts

File: src/modules/wes/hooks.js
Comment (v-rocheleau) [inline src/modules/wes/hooks.js:1]: Make new files .ts

File: src/modules/services/hooks.js
Comment (v-rocheleau) [inline src/modules/services/hooks.js:1]: Make new files .ts

File: src/modules/notifications/hooks.js
Comment (v-rocheleau) [inline src/modules/notifications/hooks.js:1]: Make new files .ts

**Review by v-rocheleau — CHANGES_REQUESTED** [review]
first pass

File: nginx.conf
Comment (davidlougheed) [reply nginx.conf:14]: no, just read up on some best practices for static file serving.

File: src/modules/metadata/hooks.js
Comment (davidlougheed) [reply src/modules/metadata/hooks.js:1]: won't bother for anything needing dispatch/state typing

File: src/components/SiteHeader.js
Comment (gsfk) [inline src/components/SiteHeader.js:103]: not sure `disabled` is doing anything here:
- managerPermissions.canManageAnything is always true if we made it to this code, so right side of disjunction is always false.
- I'm guessing left side is always false for the same reasons

File: src/components/SiteHeader.js
Comment (gsfk) [inline src/components/SiteHeader.js:112]: same reasoning here as above, will never be `disabled` since this menu item only appears conditionally

**Review by gsfk — COMMENTED** [review]
Working well will only a few issues, some may be permissions issues rather than specific to this pr. 


Regarding the site header menu items, the display is a bit inconsistent about whether you're allowed to use that menu item or not:

- always appears even when inaccessible (Overview)
- greyed out when inacessible? (Explorer?)
- shows a loading gif, then the item appears or doesn't appear at all (Data Manager / Services) .. a few inline comments below on this one. 



On testing, I got a few behaviours that are possibly intentional, and others than probably aren't:

When completely logged out: 

- I can modify the settings (the control for pie chart threshold) ... looks like this has been possible for a while. 
- I can see but not alter reference genomes
- I can't access Overview or Explorer, but "Explorer" in the menu is greyed out, while "Overview" isn't


I created a new user in Keycloak and logged in, but without managing permissions at all in authz (not sure which behaviour this is supposed to produce)

- "Explorer" is greyed out in the top menu, but I was still able to use the Explorer page, as well as click-to-search from the front page.

File: src/components/SiteHeader.js
Comment (v-rocheleau) [inline src/components/SiteHeader.js:210]: Missing: `openIdConfigFetching`, `isHandingOffCodeForToken`, `isFetchingPermissions`, `canViewNotifications`, `hasAttemptedQueryPermissions`, `hasAttemptedManagerPermissions`.

We should probably add the eslint rules for react hooks and dependency array, this way the linter can do these checks for us.
"react-hooks/rules-of-hooks": "error", // For checking rules of hooks
"react-hooks/exhaustive-deps": "warn" // For checking hook dependencies

File: src/components/explorer/SearchSummaryModal.js
Comment (v-rocheleau) [inline src/components/explorer/SearchSummaryModal.js:55]: Not related to this PR. The `katsuUrl` const is not in the useEffect's dependencies array, we probably want to check that `katsuUrl` is truthy before doing the fetch as well.

File: src/components/manager/ManagerAnalysisContent.js
Comment (v-rocheleau) [inline src/components/manager/ManagerAnalysisContent.js:27]: Can be simplified with `useHasResourcePermissionWrapper(RESOURCE_EVERYTHING, analyzeData)`.

File: src/components/manager/ManagerExportContent.js
Comment (v-rocheleau) [inline src/components/manager/ManagerExportContent.js:27]: Can be simplified with `useHasResourcePermissionWrapper(RESOURCE_EVERYTHING, analyzeData)`.

File: src/components/manager/ManagerIngestionContent.js
Comment (v-rocheleau) [inline src/components/manager/ManagerIngestionContent.js:30]: Suggestion: Here I feel a hook like `useHasResourcePermissions(resource: Resource, permissions: string[], checkAll=true)` would be useful. If `checkAll` is true, check that all the permissions are present, if it is false, check that at least one permission is present.

**Review by v-rocheleau — CHANGES_REQUESTED** [review]
Hooks suggestions

File: src/components/SiteHeader.js
Comment (davidlougheed) [reply src/components/SiteHeader.js]: good idea - i'll make a separate PR after this for that I think

File: src/components/explorer/SearchSummaryModal.js
Comment (davidlougheed) [reply src/components/explorer/SearchSummaryModal.js:55]: same as above, good point

File: src/components/manager/ManagerAnalysisContent.js
Comment (davidlougheed) [reply src/components/manager/ManagerAnalysisContent.js:27]: doesn't include a `hasAttempted` attr, so it doesn't work for this case.

File: src/components/manager/ManagerExportContent.js
Comment (davidlougheed) [reply src/components/manager/ManagerExportContent.js:27]: same here

File: src/components/manager/ManagerAnalysisContent.js
Comment (davidlougheed) [reply src/components/manager/ManagerAnalysisContent.js:27]: (should do in bento auth js to support this)

Comment (v-rocheleau) [reply src/components/manager/ManagerAnalysisContent.js:27]: Ah true, if we add `hasAttempted` we could even wrap this in an authz component RequirePermissions, similar to RequireAuth and AutoAuthenticate.

File: src/components/manager/ManagerIngestionContent.js
Comment (davidlougheed) [reply src/components/manager/ManagerIngestionContent.js:30]: I think I'll do that in a future PR as well, if that works

File: src/utils/actions.js
Comment (v-rocheleau) [inline src/utils/actions.js:73]: great idea

File: src/modules/authz/hooks.js
Comment (v-rocheleau) [inline src/modules/authz/hooks.js:114]: Maybe memoize this? This way we can skip all the `includes` operations if the permissions don't change.

File: src/modules/manager/actions.js
Comment (v-rocheleau) [inline src/modules/manager/actions.js:34]: What is this condition for?

File: src/modules/manager/actions.js
Comment (v-rocheleau) [inline src/modules/manager/actions.js:41]: Is the "OrFail" suffix still true here?

**Review by v-rocheleau — CHANGES_REQUESTED** [review]

File: src/modules/manager/actions.js
Comment (davidlougheed) [reply src/modules/manager/actions.js:34]: it's another pattern of forcing a drop box (or other state) reload - instead of calling a non-checked fetch action version, we mark the state as invalidated and then whenever `fetchIfNeeded` or whatever is called, it will re-perform the fetch since the state is marked as invalid.

File: src/modules/manager/actions.js
Comment (davidlougheed) [reply src/modules/manager/actions.js]: no, fixed

File: src/modules/manager/actions.js
Comment (davidlougheed) [reply src/modules/manager/actions.js:34]: is this fine?

**Comment by gsfk** [comment]
Still have a question left from last time: is it expected to get query data permissions when creating a new user in Keycloak (and  doing nothing else)?

**Comment by davidlougheed** [comment]
no, users shouldn't auto-get `query:data` unless there's something (an iss/client grant perhaps) giving it to them

explorer back end still works due to katsu authz not being implemented yet I think, but i'll double check the front end.

**Comment by gsfk** [comment]
With the same no-rights user I can also hand-type the /services url and get a working services page, and was able to call katsu `/individuals`...

**Comment by davidlougheed** [comment]
both are expected, services don't need permissions and katsu only checks for the no-permissions user right now because no authz yet.

**Comment by davidlougheed** [comment]
(thus why this is only still a part 1...)

**Comment by gsfk** [comment]
ah, okay, just seemed odd that services is shown as unavailable, but is still there

**Comment by davidlougheed** [comment]
it's essentially because it's not that interesting for non-managing users, but we can always make it available.

**Comment by gsfk** [comment]
I have all permissions and I'm reloading the page: what's the second spinner for? 

Peek 2024-05-08 10-21

**Review by gsfk — APPROVED** [review]
No new comments except something we've talked about already: pie charts probably shouldn't be clickable at all if the user doesn't have query data permissions (search doesn't work for them now, but they're still encouraged to click, which takes them to a "forbidden" notice).

File: src/components/manager/projects/Project.js
Comment (v-rocheleau) [inline src/components/manager/projects/Project.js:58]: Minor: wrap in a useMemo

File: src/modules/authz/hooks.js
Comment (v-rocheleau) [inline src/modules/authz/hooks.js:89]: Should `canQueryData` be a part of the `canManageAnything` permission? I get that it is not a "manager" permission per-se, but it is used in the manager to enable/disable the DRS tab, and `canManageAnything` exposes the data manager.

**Review by v-rocheleau — CHANGES_REQUESTED** [review]
Looking good, only 2 questions before approving.

File: src/modules/authz/hooks.js
Comment (davidlougheed) [reply src/modules/authz/hooks.js:89]: the problem with it being part of `canManageAnything` is that then users which can just query data (i.e., most users) can then see the manager tab... maybe in the future we can have an explicit permission for seeing the manager tab that's separate, but as long as it's derived from permissions it makes sense to leave it out, I think

Comment (v-rocheleau) [reply src/modules/authz/hooks.js:89]: Yeah I agree with you. That or we can have a permission for DRS querying and remove canQueryData from the manager permissions.

**Review by v-rocheleau — APPROVED** [review]
LGTM! Though we should wait for part 2 to be merged in `feat/manager/auth-aware` before we merge in master. Starting to review part 2 now!

**Comment by davidlougheed** [comment]
@v-rocheleau it may make it easier to work on other web branches in parallel if we merge this now and merge part 2 separately?

**Pull request merged**
